nucypher/blockchain/eth/agents.py

```diff
@@ -1525,3 +1503,44 @@ def execute(self,
         contract_function = self.contract.functions.execute(v, r, s, destination, value, data)
         receipt = self.blockchain.send_transaction(contract_function=contract_function, sender_address=sender_address)
         return receipt
+
+
+class WeightedSampler:
+    """
+    Samples random elements with probabilities proportioinal to given weights.
+    """
+
+    def __init__(self, elements: Iterable, weights: Iterable[int]):
+        assert len(elements) == len(weights)
+        self.totals = list(accumulate(weights))
+        self.elements = elements
+
+    def sample(self, rng, quantity: int, no_replacement: bool=False) -> list:
+        """
+        If ``no_replacement`` is ``False``, samples ``quantity`` of elements from the internal array.
+        Each sample is independent, and the probablity of an element to appear is proportional
+        to the weight provided to the constructor.
+
+        If ``no_replacement`` is ``True``, the elements will not repeat;
+        every time an element is sampled its weight is set to 0.
+        (does not mutate the object and only applies to the current invocation of ``sample()``).
```

Comment:
`no_replacement=False` goes against some core design principles of the network, like that no Ursula should have 2 kfrags for the same policy. So, in order to make our code as misusage-resistant as possible, I think you should simply remove the possibility of sampling with replacement. 
Also, there's the discussion about the use of "negative" names, which sometimes make APIs more complicated.

Reply:
Personally, I am a proponent of the guideline that any optional boolean parameter must be designed in a way that its default value is `False`. But I agree that since we don't need sampling with replacement, we shouldn't expose it. Removed the option.